### PR TITLE
Add `dict` methods to `ParameterDict`

### LINKED ISSUE #68476: ParameterDict should implement `.get` and `.setdefault`
## 🚀 Feature

`ParameterDict` implements most, but not all methods from `dict`. In particular it doesn't have:

    >>> set(dir({})) - set(dir(nn.ParameterDict()))
    {'__class_getitem__',
     '__ior__',
     '__or__',
     '__reversed__',
     '__ror__',
     'copy',
     'fromkeys',
     'get',
     'popitem',
     'setdefault'}

Arguably `copy`, `fromkeys` and `popitem` are not particularly useful, as well as the various `or` methods.

However `.get()` and `.setdefault()` can make code much nicer.

## Motivation

I'm implementing an optimizable optimizer with parameter specific moment and velocity.
The code looks something like this:

    def __init__(self):
        self.cache = nn.ParameterDict()
        ...
    def adjust(self, params):
        ...
        for name, param in params:
            m = self.cache.get('m', torch.zeros(param.shape))
            v = self.cache.get('v', torch.zeros(param.shape) + self.eps)
            ...

I can't define these parameter specific parameters in my `__init__`, since I only know the shape they need once the first optimization step has started.

## Pitch

The documentation writes "ParameterDict is an ordered dictionary".
It seems that as such it should implement the standard python dictionary methods.
In fact, it already implements nearly all of them, so why not those last two?

## Alternatives

I could write

    if 'm' not in self.cache:
            self.cache[name]['m'] = torch.zeros(param.shape)
    if 'v' not in self.cache 
            self.cache[name]['v'] = torch.zeros(param.shape) + self.eps
    m = self.cache[name]['m']
    v = self.cache[name]['v']

But it is 3 times as many lines of code, and any python developer would wonder "why doesn't this just use `.get` or `.default`?"

cc @albanD @mruberry @jbschlosser @walterddr

### PR DESCRIPTION
**Pull request by thisisalbertliang**
Fixes #68476

We implemented all of the following `dict` methods for `ParameterDict`
- `get `
- `setdefault`
- `popitem`
- `fromkeys`
- `copy`
- `__or__`
- `__ior__`
- `__reversed__`
- `__ror__`

The behavior of these new methods matches the expected behavior of python `dict` as defined by the language itself: https://docs.python.org/3/library/stdtypes.html#typesmapping

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -1834,11 +1834,118 @@ def check():
...
+        p_popitem = parameter_dict.popitem()
+        self.assertEqual(p_popitem[0], 'p12')
+        self.assertIs(p_popitem[1], p)
+
+        parameters['p11'] = Parameter(torch.randn(10, 10))
+        p_setdefault = parameter_dict.setdefault('p11', parameters['p11'])
+        self.assertIs(p_setdefault, parameters['p11'])
+
+        p = Parameter(torch.randn(10, 10))
+        self.assertFalse(parameter_dict.setdefault('p11', p) is p)
```
Comment (jbschlosser) [inline test/test_nn.py:1876]: Can you please also test the case where no explicit default is passed to `setdefault()`?

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Additions look correct to me and well-tested :)

Minor thing: the `third_party/tensorpipe` changes shouldn't be part of this PR - can you please remove those?

**Comment by gramliu** [comment]
@jbschlosser We addressed your comments. However, for some reason, it appears to be failing a test that's completely unrelated to the changes we made. Could we ask for help resolving this issue?

**Comment by gramliu** [comment]
Ok looks like CI is passing now

**Review by jbschlosser — APPROVED** [review]
LGTM thanks!

**Comment by bryant1410** [comment]
One more thing, just seeing this after opening #70095, shouldn't `pop` support a `default` arg as well?

**Comment by jbschlosser** [comment]
> One more thing, just seeing this after opening #70095, shouldn't `pop` support a `default` arg as well?

Hm do you mind if we repurpose #70095 to ask for that specifically and address it in a separate PR?

**Comment by bryant1410** [comment]
> @bryant1410 Hm do you mind if we repurpose #70095 to ask for that specifically and address it in a separate PR?

Okay!

**Pull request closed without merging**